Fix Crash When input_size Is an int

abduelhamit opened the pull request.

rthadur commented:
@abduelhamit please include description,what issue or bug this will be resolving.

abduelhamit commented:
Sure. Sorry, my mistake.

**System information**
- OS Platform and Distribution: Windows 10 x64 1809 (17763.805)
- TensorFlow installed from: pip
- TensorFlow version: v2.0.0-rc2-26-g64c3d382ca 2.0.0
- Python version: 3.7.4
- CUDA/cuDNN version: 10.0.130/7.6.3.30
- GPU model and memory: NVIDIA GeForce GTX 980 4GB

**Code to reproduce the issue**
I followed [tensorflow/lite/experimental/examples/lstm/TensorFlowLite_LSTM_Keras_Tutorial.ipynb](https://colab.research.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/examples/lstm/TensorFlowLite_LSTM_Keras_Tutorial.ipynb), added `tf.compat.v1` where necessary, and added `tf.compat.v1.disable_eager_execution()`.

**Other info**
With this patch, I've followed the usage of `isinstance` in [this part of the code](https://github.com/tensorflow/tensorflow/blob/1059514f7c87fc4d7afe211349fa9d0d4ac30e3d/tensorflow/lite/experimental/examples/lstm/rnn_cell.py#L301):
```python
    if len(inputs_shape) != 2:
      raise ValueError(
          "inputs_shape must be 2-dimensional, saw shape: %s" % inputs_shape)
    input_depth = (
        inputs_shape[1]
        if isinstance(inputs_shape[1], int) else inputs_shape[1].value)
    if input_depth is None:
      raise ValueError("Invalid inputs_shape, saw shape: %s" % inputs_shape)
```

**Log**
```
2019-10-24 14:54:34.023448: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
WARNING:tensorflow:From C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\ops\resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.
Instructions for updating:
If using Keras pass *_constraint arguments to layers.
Traceback (most recent call last):
  File ".\tf_lite_rnn_lstm.py", line 42, in <module>
    tf.keras.layers.Dense(10, activation=tf.nn.softmax, name='output')
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\training\tracking\base.py", line 457, in _method_wrapper
    result = method(self, *args, **kwargs)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\keras\engine\sequential.py", line 114, in __init__
    self.add(layer)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\training\tracking\base.py", line 457, in _method_wrapper
    result = method(self, *args, **kwargs)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\keras\engine\sequential.py", line 196, in add
    output_tensor = layer(self.outputs[0])
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\keras\engine\base_layer.py", line 847, in __call__
    outputs = call_fn(cast_inputs, *args, **kwargs)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\keras\layers\core.py", line 795, in call
    return self.function(inputs, **arguments)
  File ".\tf_lite_rnn_lstm.py", line 33, in buildLstmLayer
    time_major=True)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\lite\experimental\examples\lstm\rnn.py", line 266, in dynamic_rnn
    dtype=dtype)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\ops\rnn.py", line 916, in _dynamic_rnn_loop
    swap_memory=swap_memory)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\ops\control_flow_ops.py", line 2675, in while_loop
    back_prop=back_prop)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\ops\while_v2.py", line 198, in while_loop
    add_control_dependencies=add_control_dependencies)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\framework\func_graph.py", line 915, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\ops\while_v2.py", line 176, in wrapped_body
    outputs = body(*_pack_sequence_as(orig_loop_vars, args))
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\ops\rnn.py", line 884, in _time_step
    (output, new_state) = call_cell()
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\ops\rnn.py", line 870, in <lambda>
    call_cell = lambda: cell(input_t, state)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\keras\engine\base_layer.py", line 847, in __call__
    outputs = call_fn(cast_inputs, *args, **kwargs)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\python\keras\layers\recurrent.py", line 137, in call
    inputs, states = cell.call(inputs, states, **kwargs)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\site-packages\tensorflow_core\lite\experimental\examples\lstm\rnn_cell.py", line 440, in call
    if input_size.value is None:
AttributeError: 'int' object has no attribute 'value'
```

renjie-liu commented on tensorflow/lite/experimental/examples/lstm/rnn_cell.py:
```diff
@@ -439,7 +439,7 @@ def call(self, inputs, state):
 
     input_size = inputs.shape.with_rank(2)[1]
-    if input_size.value is None:
+    if not isinstance(input_size, int) and input_size.value is None:
```
thanks for the commit, how about we do:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell_impl.py#L1022-L1024

abduelhamit replied:
Yeah, that works fine. Should I change it in my patch or are you going to use that code instead?

renjie-liu replied:
can you change that? thanks a lot!

abduelhamit replied:
Done.

abduelhamit committed:
Fix Crash When input_size Is an int

rthadur commented:
Changes are merged internally waiting for auto-merge to happen.
